types/src/duration.rs

```diff
@@ -0,0 +1,317 @@
+            (1, digits)
+        };
+        let mut split = digits.splitn(2, '.');
+        let (s, n) = (split.next(), split.next());
+        let seconds = s
+            .map(str::parse::<i64>)
+            .transpose()
+            .map_err(Error::custom)?;
+        let nanos = n
+            .map(str::parse::<i32>)
+            .transpose()
+            .map_err(Error::custom)?;
+
+        Ok(Duration {
+            seconds: sign * seconds.unwrap_or(0),
+            nanos: sign as i32 * nanos.unwrap_or(0),
+        })
+    }
+}
+
+/// Implement [`serde`](::serde) deserialization for timestamps.
+impl<'de> serde::de::Deserialize<'de> for Duration {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_str(DurationVisitor)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use serde_json::json;
+    use test_case::test_case;
+
+    // Verify 0 converts as expected.
+    #[test]
+    fn zero() {
+        let proto = Duration {
+            seconds: 0,
+            nanos: 0,
+        };
+        let json = serde_json::to_value(&proto).unwrap();
+        let expected = json!(r#"0s"#);
+        assert_eq!(json, expected);
+        let roundtrip = serde_json::from_value::<Duration>(json).unwrap();
+        assert_eq!(proto, roundtrip);
+    }
+
+    // Verify 0 converts as expected.
+    #[test_case(0, 0, 0, 0 ; "all inputs are zero")]
+    #[test_case(1, 0, 1, 0 ; "positive seconds and zero nanos")]
+    #[test_case(1, 200_000, 1, 200_000 ; "positive seconds and nanos")]
+    #[test_case(-1, 0, -1, 0; "negative seconds and zero nanos")]
+    #[test_case(-1, -500_000_000, -1, -500_000_000; "negative seconds and nanos")]
+    #[test_case(2, -400_000_000, 1, 600_000_000; "positive seconds and negative nanos")]
+    #[test_case(-2, 400_000_000, -1, -600_000_000; "negative seconds and positive nanos")]
+    fn normalize(seconds: i64, nanos: i32, want_seconds: i64, want_nanos: i32) {
+        let got = Duration::new(seconds, nanos);
+        let want = Duration {
+            seconds: want_seconds,
+            nanos: want_nanos,
+        };
+        assert_eq!(want, got);
+    }
+
+    #[test_case(0, 0, 0; "zero nanos")]
+    #[test_case(1_000_000_000, 1, 0; "one second")]
+    #[test_case(1_200_000_000, 1, 200_000_000; "one second and 200ms")]
+    #[test_case(-1_200_000_000, -1, -200_000_000; "minus one second and 200ms")]
+    fn from_nanos(nanos: i32, want_seconds: i64, want_nanos: i32) {
+        let got = Duration::from_nanos(nanos);
+        let want = Duration {
+            seconds: want_seconds,
+            nanos: want_nanos,
+        };
+        assert_eq!(want, got);
+    }
+
+    #[test_case(0, 0, "0s" ; "zero")]
+    #[test_case(12, 0, "12s"; "round positive seconds")]
+    #[test_case(12, 123, "12.000000123s"; "positive seconds and nanos")]
+    #[test_case(12, 123_000, "12.000123000s"; "positive seconds and micros")]
```

Comment:
Here, and a few cases below, should we be instead expecting no trailing zeros? `12.000123000s` => `12.000123s`. I don't think it's incorrect to have trailing `0`s but suspect you'd want to encode as `12.1s` instead of `12.100000000s`.

Reply:
Maybe, I think both are legal encodings, though the trailing zeroes are ugly. Let me open a bug.